src/kernel/panic.zig

```diff
@@ -1,13 +1,148 @@
+const std = @import("std");
 const builtin = @import("builtin");
 const tty = @import("tty.zig");
 const arch = @import("arch.zig").internals;
 const log = @import("log.zig");
+const multiboot = @import("multiboot.zig");
+const mem = @import("mem.zig");
+const ArrayList = std.ArrayList;
+const expectEqual = std.testing.expectEqual;
+
+var symbol_map: ?SymbolMap = null;
+
+const MapEntry = struct {
+    addr: usize,
+    func_name: []const u8,
+};
+
+const SymbolMap = struct {
+    symbols: ArrayList(MapEntry),
+
+    pub fn init(allocator: *std.mem.Allocator) SymbolMap {
+        return SymbolMap{
+            .symbols = ArrayList(MapEntry).init(allocator),
+        };
+    }
+
+    pub fn add(self: *SymbolMap, name: []const u8, addr: u32) !void {
+        try self.symbols.append(MapEntry{ .addr = addr, .func_name = name });
+    }
+
+    pub fn search(self: *const SymbolMap, addr: usize) ?[]const u8 {
+        if (self.symbols.count() == 0)
+            return null;
+        // Find the first element whose address is greater than addr
+        var previous_name: ?[]const u8 = null;
+        for (self.symbols.toSlice()[0..]) |entry| {
+            if (entry.addr > addr)
+                return previous_name;
+            previous_name = entry.func_name;
+        }
+        return previous_name;
+    }
+};
+
+pub fn init(mem_profile: *const mem.MemProfile, allocator: *std.mem.Allocator) !void {
+    log.logInfo("Init panic\n");
+    // Exit if we haven't loaded all debug modules
+    if (mem_profile.boot_modules.len < 1)
+        return;
+    var kmap_start: u32 = 0;
+    var kmap_end: u32 = 0;
+    for (mem_profile.boot_modules) |module| {
+        const mod_start = mem.physToVirt(module.mod_start);
+        const mod_end = mem.physToVirt(module.mod_end);
+        const mod_str_ptr = mem.physToVirt(@intToPtr([*]u8, module.cmdline));
+        const mod_str = mod_str_ptr[0..strlen(mod_str_ptr)];
+        if (std.mem.eql(u8, mod_str, "kernel.map")) {
+            kmap_start = mod_start;
+            kmap_end = mod_end;
+            break;
+        }
+    }
+    std.debug.assert(kmap_start != 0 and kmap_end != 0);
+
+    var syms = SymbolMap.init(allocator);
```

Comment:
Could have an `errdefer` to deinit the table if there was an error creating an entry. idk if you need all entries to perform a back trace. Or, lets say an entry halfway through is corrupt, could the loop below continue with the other entries and just insert a null for missing ones?

Reply:
We do need all entries for the map to be valid so I'll make an `errdefer` and make the loop return a `InvalidSymbolMap` error if it is corrupt.